### DISCUSSION TITLE
How to add new file based on another files contents using the web interface?

### DISCUSSION THREAD

**Discussion by kamilkrzyskow** [community/community#46462] [Repositories] [answered]
### Select Topic Area

Question

## Context
For context I would like to simplify the process of adding localized files to a repository containing a documentation. It's based on MkDocs with the Material for MkDocs theme, and the i18n plugin for MkDocs. The official site is hosted on GitHub pages and a development site is hosted externally.

### Solution attempt 1
I have a button (provided by the theme) which allows to edit the raw `index.md` file, the link ends like this:
```
.../edit/dev/docs/index.md
```
adding the `?filename=index.lang.md` param (lang being a 2 character language identifier) to the link only changes the name of the edited file and in result the commit deletes the previous file. A param like `createnew=1` would be perfect, but I didn't find anything similar looking in the files within the browser Developer Tools.

### Solution attempt 2
So another approach is to use a link which creates a new file with the contents of the previous file:
```
.../new/dev/docs/?filename=index.lang.md&value=<CONTENTS>
```
the issue here is that it only works for smaller files, and a longer file from the repository resulted in a `Whoa there! Your request URL is too long.` error, and another issue is that currently I'm planning to fetch the raw contents dynamically after button click (otherwise I would have to include a URL encoded site during build time), and CORS is blocking it from the external development site (minor issue, I'm just pointing that out).

## Any Ideas?
After those 2 attempts, I feel like I need to scrap the idea of the 1 click button which adds a file with another files contents and will have to use the second attempt with a message `.../new/dev/docs/?filename=index.lang.md&value=Copy the contents of the original file yourself ;)`

**Comment by 2MuchC0ff33** [comment]
Thank you for your question @kamilkrzyskow 

One option is to use the GitHub API to create a new file from the contents of another file. To create a new file, send a POST request to the API endpoint and include the contents in the request body. This would allow you to retrieve the original file's contents dynamically and include them in the API request without worrying about URL length limits. Furthermore, the API does not have the same restrictions as the web interface, so you will not encounter CORS issues.

Here's an example in Bash of how to use the GitHub API v3:
```
POST /repos/{owner}/{repo}/contents/{path}

{
  "message": "create index.lang.md",
  "content": "base64 encoded contents of original file",
  "branch": "dev"
}
```

Replace owner, repo, and path with the values appropriate for your repository. The original file's base64 encoded contents should be stored in the content field. The branch field specifies which branch the file should be created in.

You can use a library or an HTTP client in your preferred programming language to make this API call.

**Reply by kamilkrzyskow** [reply]
I guess, I didn't consider using the API to aid this process. I will have to think if I can incorporate it in my final solution. 🤔 

I wanted to use the web interface approach, because it allows to make changes to the file and automatically creates a new fork repository for external users, after they save their changes and doesn't commit anything when they cancel. This would simplify the contribution process for translators since they'd not need to run any external software (aka git cli / python project setup etc.).

Your suggestion allows to create a new file, that is a exact copy of the file, and I could then allow the users to edit those files on a external branch, but if they'd cancel their changes it would still leave that branch hanging. Also I don't know how safe / if it's possible to keep it safe, would be exposing a API hook in the website button that sends POST requests, that creates a new branch with a new file. I'm also not sure how to properly guide such user to make a PR after the change 🤔 

Many things to consider, I will mark it as an answer later if I decide to use any API requests.

**Reply by 2MuchC0ff33** [reply]
As I understand, you'd like to keep the contribution process as simple as possible for translators. In that case, consider creating a template file containing the boilerplate for a new localisation file and then instruct translators on creating a new file based on the template. They would only need to copy the template file and make their changes this way, eliminating the need for API calls or external tools.

If you decide to expose the API hook, keeping security in mind is critical. Please consider using proper authentication and authorisation mechanisms to ensure that only authorised users can make API calls to create new branches and files. Also, clear documentation and guidance on using the API and making changes to the repository via the API may be helpful.

**Reply by kamilkrzyskow** [reply, marked as answer]
Sorry for the late response. I was supposed to write it sooner, but things got in the way, and then I forgot :sweat_smile:.

In the end I didn't figure out a safe, and most of all straightforward way for random people to be able to add translations via some tool with the API. It needs proper authentication to not be abused, also as far as I know people would have to get their own personal token and/or get a copy of the files etc.

Too many points of friction, so ultimately I didn't choose it. As a side-note, looking over the API docs, the endpoint you've provided used the `PUT` method instead of `POST`, however I didn't actually try checking to see how it would behave.

I went instead with the idea I had when I started this discussion, and you've also alluded to it in your last reply. I fill out the `/new/` link data with instructions for the user to follow, which include copying the file and replacing those instructions.

At first, it worked fine, then the new GitHub UI was introduced, and [this approach got "corrupted"](https://github.com/orgs/community/discussions/48301#discussioncomment-5199609). I still think that, even though it has "a lot of" explanatory steps, the user experience basically comes down to:

1. Click the (+) icon (GitHub opens),
2. Read the instructions (once),
3. Copy the raw original English text (from a another browser tab),
4. Replace the instructions with the copied text,
5. Translate and commit.

Therefore, it solves the problem stated in this issue's title, as everything is happening via the GitHub web interface.

---

As for the actual solution it is split up into 3 different parts:

1. The mkdocs i18n plugin provides information about the locale of a given page. I pass this information into the client-side JavaScript.
   
    https://github.com/Gothic-Modding-Community/gmc/blob/812bef0ba16cd4690bac8a6d759fe7fe6f00944f/overrides/main.html#L45-L51

2. Based on this information, when I detect that the chosen language is different from English, and the page is still displayed in English I add a button to prompt the user to translate.
   
    https://github.com/Gothic-Modding-Community/gmc/blob/16ef42ad3cba01974d33850d2e209ad144ce1da0/overrides/assets/javascripts/extra.js#L103-L105

3. The button is an anchor with a link with specified instructions for the user.
   
    https://github.com/Gothic-Modding-Community/gmc/blob/812bef0ba16cd4690bac8a6d759fe7fe6f00944f/overrides/assets/javascripts/extra.js#L353

This single page likely won't ever be translated, so anyone interested can see the example in [action here](https://gothic-modding-community.github.io/gmc/cs/contribute/#add-translated-pages).
